packages/rollup-core/src/app/data/data-service.ts

```diff
@@ -452,6 +457,43 @@ export class DefaultDataService implements DataService {
     }
   }
 
+  public async getMaxL2TxOutputIdForCanonicalChainBatch(
+    batchTimestamp: number,
+    maxBatchCalldataBytes: number
+  ): Promise<number> {
+    const res: Row[] = await this.rdb.select(
+      `SELECT id, (LENGTH(calldata)-2) / 2 as calldata_bytes
+      FROM l2_tx_output
+      WHERE 
+            block_timestamp = ${batchTimestamp}
+            AND canonical_chain_batch_number IS NULL
+      ORDER BY id ASC`
+    )
+
+    if (!res || !res.length) {
+      const msg: string = `Was told to get max L2 Tx Output ID for batch timestamp ${batchTimestamp}, but no records found!`
+      log.error(msg)
+      throw Error(msg)
+    }
+
+    let totalCalldataBytes: number = 0
+    let lastId = -1
+    for (const row of res) {
+      totalCalldataBytes += parseInt(row['calldata_bytes'], 10)
+      if (totalCalldataBytes > maxBatchCalldataBytes) {
```

Comment:
This codepath wouldn't be impacted from a ton of transactions being included in a batch? The batch size would be selected by the sequencer? So users submitting tons of txs wouldn't be able to trigger this?

Reply:
It's possible that we'd handle a ton of txs with the same block timestamp, at which point this collection would be very large, but we'll eagerly exit this loop as soon as our max gets hit, so we should be good. Might be good to add an arbitrary [but large] limit to the query, though, just in case we have a burst of txs with 0 calldata.

Reply:
Worth noting, I created a query that does this calc in the query itself, but it's slow and would get slower as more records get added (stupid `CROSS JOIN` 😆 )